### DISCUSSION TITLE
Questions

### DISCUSSION THREAD

**Discussion by DoTulip** [carnotresearch/cr-sparse#27] [Q&A]
Hello, Dr. Kumar,
    This is really a very valuable job! I have two points that are not very clear: 1. Will cr.sparse support solving generalized eigenvalue problems in the future? 2. When you use BCOO, how do you define the automatic differentiation rules for related operations? I am looking forward to your reply, thank you.

**Comment by shailesh1729** [comment]
@DoTulip 

Hi, 

- I don't have a definitive answer on the support for solving generalized eigenvalue problems at the moment. Currently, I am working on porting the [tfocs](http://cvxr.com/tfocs/) library-related stuff.  I build support for relevant numerical linear algebra subroutines as and when I need them for higher-level algorithms. However, contributions are welcome. 
- BCOO is a storage format for sparse matrices provided by the JAX development team. as per their [documentation](https://jax.readthedocs.io/en/latest/jax.experimental.sparse.html), they are doing extensive work for supporting all JAX related infrastructure. I believe they would have full support for automatic differentiation too. You may want to ask on their discussion forum for details.
- For my own work in CR-Sparse, I haven't used the auto-diff capabilities of JAX much. 

Do let me know what kind of things you are looking to use this library for.  

BTW, I have recently separated out (refactored) linear algebra-related stuff into a separate library [cr-nimble](https://github.com/carnotresearch/cr-nimble).  This way, cr-nimble will have more general-purpose use for people who are just interested in things like truncated svd, eig, etc. while cr-sparse will continue to focus on sparse recovery-related problems.
